services/galley/src/Galley/API/MLS/Message.hs

```diff
@@ -179,19 +179,45 @@ postMLSMessage ::
   ) =>
   Local x ->
   Qualified UserId ->
+  Maybe ConvId ->
   Maybe ConnId ->
   RawMLS SomeMessage ->
   Sem r [LocalConversationUpdate]
-postMLSMessage loc qusr con smsg = case rmValue smsg of
+postMLSMessage loc qusr convId con smsg = case rmValue smsg of
   SomeMessage _ msg -> do
     -- fetch conversation ID
     qcnv <- getConversationIdByGroupId (msgGroupId msg) >>= noteS @'ConvNotFound
+    when (maybe False (qUnqualified qcnv /=) convId) $ throwS @'MLSUnsupportedMessage
+    unless (msgEpoch msg == Epoch 0) $
+      flip unless (throwS @'MLSUnsupportedMessage) =<< isUserSender qusr smsg
     foldQualified
       loc
       (postMLSMessageToLocalConv qusr con smsg)
       (postMLSMessageToRemoteConv loc qusr con smsg)
       qcnv
 
+isUserSender ::
```

Comment:
I'm unsure of what exactly is being tested here. We only care about the equalities for one specific use case, and default to `True` otherwise?

Reply:
It's checking that the MLS client who created the message belongs to the user who is the sender of the REST request, identified by HTTP header. This is only relevant in an ongoing conversation. The check should be skipped in case of 
* `SMLSCipherText`: encrypted messages in which we don't have access to the sending client's key package, 
* `PreconfiguredSender`: messages sent by the backend, and
* `NewMemberSender`: external add proposals which propose fresh key packages for new clients and thus the validity of the key package cannot be known at the time of this check.

External add proposals are validated at a later stage of message processing and so will be messages sent by the backend. The latter is not in place yet, I think. There's not much the backend can do with/about encrypted messages other than passing them forward to the clients (and perhaps query the clients for news).

Reply:
Thanks, that's much clearer. Would it be good to add this summary as a comment on top of the function? Or point to the spec/some doc that clarifies this?

Reply:
Yep, realised after writing my reply that a comment in code could do no harm to this function.